File: hash-table/subarray-sum-equals-k/subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
function subarraySum(nums, k) {
  let sum = 0;
  const prefix = nums.map(el => {
    sum += el;
    return sum;
  })

  const sumsMap = {};
  return nums.reduce((acc, curr, index) => {
    if(prefix[index] === k) {
      acc++;
    }
    
    if (sumsMap[prefix[index] - k]) {
      acc += sumsMap[prefix[index] - k];
    }
    
    sumsMap[prefix[index]] =(sumsMap[prefix[index]] || 0) + 1;
    return acc;

  }, 0);
}

File: hash-table/subarray-sum-equals-k/subarray-sum-equals-k.go
func subarraySum(nums []int, k int) int {
	prefix := make([]int, len(nums))
	prefix[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		prefix[i] = nums[i] + prefix[i-1]
	}

	ans := 0
	sum := make(map[int]int)
	for _, p := range prefix {
		if p == k {
			ans++
		}

		diff := p - k

		if pp, ok := sum[diff]; ok {
			ans += pp
		}

		sum[p]++
	}

	return ans
}
